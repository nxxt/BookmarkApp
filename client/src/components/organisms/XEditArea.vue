<template>
  <v-container fluid class="pa-1 ma-0" @click="$emit('edit-area-click')">
    <v-row>
      <v-col class="d-flex align-center">
        <v-tooltip bottom>
          <template v-slot:activator="{ on, attrs }">
            <v-btn
              v-if="isMobile"
              icon
              small
              class="mr-1 x-edit-area-menu-btn"
              v-bind="attrs"
              v-on="on"
              @click.stop="$emit('menu-btn-click')"
            >
              <v-icon>mdi-menu</v-icon>
            </v-btn>
          </template>
          <span>toggle sidebar</span>
        </v-tooltip>

        <v-tooltip bottom>
          <template v-slot:activator="{ on, attrs }">
            <v-btn
              class="x-edit-area-arrow-left-btn"
              small
              v-bind="attrs"
              icon
              v-on="on"
              @click="$emit('arrow-left-btn-click')"
            >
              <v-icon>mdi-arrow-left</v-icon>
            </v-btn>
          </template>
          <span>back to</span>
        </v-tooltip>
      </v-col>
    </v-row>

    <v-row class="scroll">
      <v-col class="d-flex justify-center">
        <slot></slot>
      </v-col>
    </v-row>
  </v-container>
</template>

<script lang="ts">
import { Component, Prop, Vue } from 'vue-property-decorator'

@Component({})
export default class XEditArea extends Vue {
  @Prop({ type: Boolean, default: false })
  isMobile!: boolean
}
</script>
