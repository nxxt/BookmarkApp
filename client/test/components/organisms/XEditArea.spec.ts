import Vuetify from 'vuetify'
import { createLocalVue, mount } from '@vue/test-utils'
import XEditArea from '@/components/organisms/XEditArea.vue'

const localVue = createLocalVue()

describe('XEditArea.vue', () => {
  /****************
   * 準備
   ****************/
  // vuetify
  let vuetify: typeof Vuetify
  beforeEach(() => {
    vuetify = new Vuetify()
  })

  // propsで渡す値
  const propsData = {
    isMobile: false
  }

  const factory = (options?: any) => {
    return mount(XEditArea, {
      localVue,
      vuetify,
      ...options
    })
  }

  /****************
   * TEST
   ****************/
  it('props', () => {
    const wrapper = factory({ propsData })
    expect(wrapper.props()).toStrictEqual(propsData)
  })

  describe('template', () => {
    it('slot', () => {
      const wrapper = factory({
        slots: {
          default: '<div data-test="slotContent">slot content</div>'
        }
      })

      const slotContent = wrapper.find('[data-test="slotContent"]')
      expect(slotContent.exists()).toBe(true)
      expect(slotContent.text()).toBe('slot content')
    })

    it('menu btn visible', () => {
      const wrapper = factory()
      const btn = wrapper.find('.x-edit-area-menu-btn')
      expect(btn.exists()).not.toBe(true)

      wrapper.setProps({ isMobile: true })

      wrapper.vm.$nextTick(() => {
        const btn = wrapper.find('.x-edit-area-menu-btn')
        expect(btn.exists()).toBe(true)
      })
    })
  })

  describe('emit', () => {
    it('container', () => {
      const wrapper = factory()
      const btn = wrapper.find('.container')
      const event = jest.fn()
      wrapper.vm.$on('edit-area-click', event)

      // 初期状態
      expect(event).toBeDefined() // イベントがundefinedでないこと
      expect(event).toHaveBeenCalledTimes(0) // eventが発生していないこと

      // emit
      btn.trigger('click')
      // イベント後
      expect(event).toHaveBeenCalledTimes(1) // イベントが親へ伝搬したこと
      expect(wrapper.emitted('edit-area-click')).toBeDefined()
    })

    it('menu btn', () => {
      const wrapper = factory({ propsData: { isMobile: true } })
      const btn = wrapper.find('.x-edit-area-menu-btn')
      const event = jest.fn()
      wrapper.vm.$on('menu-btn-click', event)

      // 初期状態
      expect(event).toBeDefined() // イベントがundefinedでないこと
      expect(event).toHaveBeenCalledTimes(0) // eventが発生していないこと

      wrapper.vm.$nextTick(() => {
        // emit
        // × btn.vm.$emit('click')
        btn.trigger('click')
        // イベント後
        expect(event).toHaveBeenCalledTimes(1) // イベントが親へ伝搬したこと
        expect(wrapper.emitted('menu-btn-click')).toBeDefined()
      })
    })

    it('arrow-left btn', () => {
      const wrapper = factory()
      const btn = wrapper.find('.x-edit-area-arrow-left-btn')
      const event = jest.fn()
      wrapper.vm.$on('arrow-left-btn-click', event)

      // 初期状態
      expect(event).toBeDefined() // イベントがundefinedでないこと
      expect(event).toHaveBeenCalledTimes(0) // eventが発生していないこと

      // emit
      btn.trigger('click')
      // イベント後
      expect(event).toHaveBeenCalledTimes(1) // イベントが親へ伝搬したこと
      expect(wrapper.emitted('arrow-left-btn-click')).toBeDefined()
    })
  })

  describe('the entire template', () => {
    it('snapshot', () => {
      const wrapper = factory({ propsData })
      expect(wrapper.html()).toMatchSnapshot()
    })
  })
})
