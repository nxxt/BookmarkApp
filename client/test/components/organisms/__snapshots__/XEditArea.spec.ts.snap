// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`XEditArea.vue the entire template snapshot 1`] = `
"<div class=\\"container pa-1 ma-0 container--fluid\\">
  <div class=\\"row\\">
    <div class=\\"d-flex align-center col\\"><span class=\\"v-tooltip v-tooltip--bottom\\"><!----></span> <span class=\\"v-tooltip v-tooltip--bottom\\"><!----><button type=\\"button\\" class=\\"x-edit-area-arrow-left-btn v-btn v-btn--flat v-btn--icon v-btn--round theme--light v-size--small\\" role=\\"button\\" aria-haspopup=\\"true\\" aria-expanded=\\"false\\"><span class=\\"v-btn__content\\"><i aria-hidden=\\"true\\" class=\\"v-icon notranslate mdi mdi-arrow-left theme--light\\"></i></span></button></span></div>
  </div>
  <div class=\\"row scroll\\">
    <div class=\\"d-flex justify-center col\\"></div>
  </div>
</div>"
`;
